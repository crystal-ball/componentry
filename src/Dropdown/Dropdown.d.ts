import * as React from 'react'
import {
  ActiveContainerProps,
  ActiveTriggerProps,
  ActiveContentProps,
} from '../utils/componentry-props'

declare const DrodpownContent: React.ComponentType<ActiveContentProps>
declare const DropdownItem: React.ComponentType<ActiveTriggerProps>
declare const DrodpownTrigger: React.ComponentType<ActiveTriggerProps>

interface DropdownComponent extends React.ComponentClass<ActiveContainerProps> {
  Content: typeof DrodpownContent
  Item: typeof DropdownItem
  Trigger: typeof DrodpownTrigger
}

/** https://crystal-ball.github.io/componentry/components/dropdowns */
declare const Dropdown: DropdownComponent

export default Dropdown
