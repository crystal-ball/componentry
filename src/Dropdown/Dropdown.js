import activeContainer from '../active-container-factory'
import activeContent from '../active-content-factory'
import activeTrigger from '../active-trigger-factory'
import withActive from '../withActive'
import withTheme from '../withTheme'

const Content = activeContent({
  arias: { labelledby: true, hidden: true },
  classes: 'dropdown-content',
})

const Trigger = activeTrigger({
  arias: { expanded: true, haspopup: true, id: true },
  classes: 'dropdown-toggle',
})

const Item = activeTrigger({
  // TODO: what arias should this have?
  arias: {},
  classes: 'dropdown-item',
  // Dropdown items have appropriate styles
  defaultLink: false,
})

const Dropdown = activeContainer({
  Content: withTheme('DropdownContent', withActive(Content)),
  Trigger: withTheme('DropdownTrigger', withActive(Trigger)),
  classes: 'dropdown',
  escHandler: true,
  clickHandler: true,
})
Dropdown.Item = withTheme('DropdownItem', withActive(Item))
Dropdown.defaultProps = {
  ...Dropdown.defaultProps,
  direction: 'bottom',
}

/**
 * The `<Dropdown>` element creates a menu.
 *
 * TODO:
 * - Default focus on open first item
 * - Keydown listener for arrows to navigate through menu items
 */
export default withTheme('Dropdown', Dropdown)
