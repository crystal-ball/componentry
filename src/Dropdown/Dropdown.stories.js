/* global module */

import React from 'react'
import { storiesOf } from '@storybook/react'
import { select } from '@storybook/addon-knobs'
import { action } from '@storybook/addon-actions'

import Dropdown from './Dropdown'

const directions = ['top', 'right', 'bottom', 'left', 'overlay']

storiesOf('|Dropdown', module).add('<Dropdown />', () => (
  <Dropdown
    onActivate={action('activating')}
    onDeactivate={action('deactivating')}
    direction={select('direction', directions, 'bottom')}
  >
    <Dropdown.Trigger>Trigger</Dropdown.Trigger>
    <Dropdown.Content>
      <h4 className='dropdown-heading'>Available actions</h4>
      <Dropdown.Item>Interactive Item 1</Dropdown.Item>
      <Dropdown.Item>Interactive Item 2</Dropdown.Item>
      <Dropdown.Item href='https://github.com/crystal-ball/componentry'>
        Componentry
      </Dropdown.Item>
      <Dropdown.Item onClick={action('item click')} disabled>
        Disabled Item
      </Dropdown.Item>
      <div className='dropdown-divider' />
      <span className='dropdown-item-text'>Dropdown item text is not interactive</span>
    </Dropdown.Content>
  </Dropdown>
))
