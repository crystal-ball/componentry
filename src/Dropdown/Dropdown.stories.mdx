import { select } from '@storybook/addon-knobs'
import { action } from '@storybook/addon-actions'

import Dropdown from './Dropdown'
import { directions } from '../storybook-resources'

<Meta title='Components|Dropdown' component={Dropdown} />

<DocsTitle title="Dropdowns" experimental/>

<Dropdown
  onActivate={action('activating')}
  onDeactivate={action('deactivating')}
  direction={select('direction', directions, 'bottom')}
>
  <Dropdown.Trigger>Trigger</Dropdown.Trigger>
  <Dropdown.Content>
    <h4 className='dropdown-heading'>Available actions</h4>
    <Dropdown.Item>Interactive Item 1</Dropdown.Item>
    <Dropdown.Item>Interactive Item 2</Dropdown.Item>
    <Dropdown.Item href='https://github.com/crystal-ball/componentry'>
      Componentry
    </Dropdown.Item>
    <Dropdown.Item onClick={action('item click')} disabled>
      Disabled Item
    </Dropdown.Item>
    <div className='dropdown-divider' />
    <span className='dropdown-item-text'>Dropdown item text is not interactive</span>
  </Dropdown.Content>
</Dropdown>

## Variant

<Dropdown className='d-block w-50'>
  <Dropdown.Trigger
    as='input'
    type='text'
    variant='query-bar'
    placeholder='Enter search'
    className='w-100'
  />
  <Dropdown.Content>
    <Dropdown.Item>Interactive Item 1</Dropdown.Item>
    <Dropdown.Item>Interactive Item 2</Dropdown.Item>
  </Dropdown.Content>
</Dropdown>
