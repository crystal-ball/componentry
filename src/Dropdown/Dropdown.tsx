import { activeContainerBuilder } from '../utils/active-container-component-builder'
import { activeActionComponent } from '../utils/active-action-component-builder'
import { activeContentComponent } from '../utils/active-content-component-builder'
import {
  BaseActiveActionProps,
  BaseActiveContainerProps,
  BaseActiveContentProps,
  BaseProps,
} from '../utils/base-types'
import { DrawerContentProps } from '../Drawer/Drawer'

export interface DropdownProps
  extends BaseActiveContainerProps,
    BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}

export interface DropdownActionProps
  extends BaseActiveActionProps,
    BaseProps,
    Omit<React.ComponentPropsWithoutRef<'button'>, 'className'> {
  /** Display variant */
  variant?: 'primary'
}

export interface DropdownContentProps
  extends BaseActiveContentProps,
    BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {
  /** Display variant */
  variant?: 'primary'
}

export interface DropdownItemProps
  extends BaseActiveActionProps,
    BaseProps,
    Omit<React.ComponentPropsWithoutRef<'button'>, 'className'> {
  /** Display variant */
  variant?: 'unstyled'
}

export interface Dropdown {
  (props: DropdownProps): React.ReactElement
  /**
   * [Dropdown action component üìù](https://componentry.design/components/dropdown)
   */
  Action: React.FC<DropdownActionProps>
  /**
   * [Dropdown content component üìù](https://componentry.design/components/dropdown)
   */
  Content: React.FC<DrawerContentProps>
  /**
   * [Dropdown item component üìù](https://componentry.design/components/dropdown)
   */
  Item: React.FC<DropdownItemProps>
}

/**
 * [Dropdown component üìù](https://componentry.design/components/dropdown)
 */
export const Dropdown = activeContainerBuilder<DropdownProps>('Dropdown', {
  clickEvents: true,
  direction: 'bottom',
  escEvents: true,
}) as Dropdown

Dropdown.Action = activeActionComponent<DropdownActionProps>('DropdownAction', {
  aria: { expanded: true, haspopup: true, id: true },
})

Dropdown.Content = activeContentComponent<DropdownContentProps>('DropdownContent', {
  aria: { labelledby: true, hidden: true },
})

Dropdown.Item = activeActionComponent<DropdownItemProps>('DropdownItem')
