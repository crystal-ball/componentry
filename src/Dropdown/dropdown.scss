/* componentry/src/Dropdown/dropdown */

//
// --- Container styles
//

// The dropdown wrapper element handles creating a relative container that the
// content is absolutely positioned off of
.dropdown {
  position: relative;
  // Default dropdown container to inline-block so that click area is constrained
  // to dropdown trigger
  display: inline-block;
}

//
// --- Trigger styles
//

.dropdown-action-primary {
  @include active-trigger-element;

  font-size: $dropdown-font-size;
  font-weight: $dropdown-font-weight;
}

//
// --- Content styles
//

.dropdown-content {
  display: block;
  position: absolute;
  z-index: $zindex-dropdown;
  min-width: $dropdown-min-width;
  padding: $dropdown-padding-y 0;
  color: theme-color(foreground);
  font-size: $dropdown-font-size;
  font-weight: $dropdown-font-weight;
  text-align: left; // Ensures proper alignment if parent has it changed (e.g., modal footer)
  background-color: $dropdown-bg;
  background-clip: padding-box;
  border: $dropdown-border-width solid $dropdown-border-color;
  @include border-radius($dropdown-border-radius);
  box-shadow: box-shadow($dropdown-box-shadow);

  // Directional styles with parent selector

  .overlay & {
    top: 0;
  }

  .top & {
    top: auto;
    bottom: 100%;
    margin-bottom: $dropdown-spacer;
  }

  .right & {
    top: 0;
    right: auto;
    left: 100%;
    margin-left: $dropdown-spacer;
  }

  .bottom & {
    top: 100%;
    left: 0;
    margin-top: $dropdown-spacer;
  }

  .left & {
    top: 0;
    right: 100%;
    left: auto;
    margin-right: $dropdown-spacer;
  }

  // Aligns dropdown content to right edge of dropdown trigger instead of left
  &.align-right {
    right: 0;
    left: auto;
  }
}

//
// Sizes
//

.dropdown-sm {
  .dropdown-content,
  .a-dropdown-trigger {
    font-size: $dropdown-font-size-sm;
    font-weight: $dropdown-font-weight-sm;
  }

  .dropdown-item {
    padding: $dropdown-item-padding-y-sm $dropdown-item-padding-x-sm;
  }
}

.dropdown-lg {
  .dropdown-content,
  .a-dropdown-trigger {
    font-size: $dropdown-font-size-lg;
    font-weight: $dropdown-font-weight-lg;
  }

  .dropdown-item {
    padding: $dropdown-item-padding-y-lg $dropdown-item-padding-x-lg;
  }
}

//
// Content items, text, headers and dividers
//

// Dropdown items
// `<button>`-specific styles are denoted with `// For <button>s`
.dropdown-item {
  display: block;
  width: 100%;
  padding: $dropdown-item-padding-y $dropdown-item-padding-x;
  // Ensure that button and anchor elements inherit dropdown font styles
  font-size: inherit;
  font-weight: $font-weight-normal;
  color: $dropdown-action-item-color;
  line-height: $dropdown-item-line-height;
  white-space: nowrap; // prevent links from randomly breaking onto new lines
  text-align: inherit; // For `<button>`s
  background-color: transparent; // For `<button>`s
  border: 0; // For `<button>`s

  &:hover,
  &:focus {
    color: $dropdown-action-item-hover-color;
    text-decoration: none;
    background: gradient-bg($dropdown-action-item-hover-bg);
  }

  &.active,
  &:active {
    color: $dropdown-active-item-color;
    text-decoration: none;
    background: gradient-bg($dropdown-active-item-bg);
  }

  &.disabled {
    color: $dropdown-disabled-item-color;
    background-color: transparent;
  }
}

// Dropdown text
.dropdown-item-text {
  display: block;
  padding: $dropdown-item-padding-y $dropdown-item-padding-x;
  color: typography-color(foreground);
}

// Dropdown section headers
.dropdown-heading {
  display: block;
  padding: $dropdown-padding-y $dropdown-item-padding-x;
  margin-top: 0;
  margin-bottom: 0;
  font-size: $font-size-sm;
  color: $dropdown-heading-color;
  white-space: nowrap; // as with > li > a
}

// Dividers (basically an `<hr>`) within the dropdown
.dropdown-divider {
  height: 0;
  margin: $space-sm 0;
  overflow: hidden;
  border-top: 1px solid $dropdown-divider-bg;
}
