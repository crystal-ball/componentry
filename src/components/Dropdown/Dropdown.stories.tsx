import type { Meta, StoryObj } from '@storybook/react'

import { Dropdown } from './Dropdown'

const meta: Meta<typeof Dropdown> = {
  component: Dropdown,
}

export default meta
type Story = StoryObj<typeof Dropdown>

export const Primary: Story = {
  args: {
    direction: 'bottom',
    children: (
      <>
        <Dropdown.Action>Action</Dropdown.Action>
        <Dropdown.Content>
          <h4 className='dropdown-heading'>Available actions</h4>
          <Dropdown.Item>Dropdown Item</Dropdown.Item>
          {/* @ts-expect-error -- experimental component */}
          <Dropdown.Item href='https://github.com/crystal-ball/componentry'>
            Link item
          </Dropdown.Item>
          <Dropdown.Item onClick={console.log}>Button item</Dropdown.Item>
          <Dropdown.Item disabled onClick={console.log}>
            Disabled button
          </Dropdown.Item>
          <div className='dropdown-divider' />
          <div className='dropdown-item-text'>Dropdown item text is not interactive</div>
        </Dropdown.Content>
      </>
    ),
  },
}

export const WithMultiple: Story = {
  args: {
    className: 'd-block w-50',
    children: (
      <>
        {/* @ts-expect-error -- experimental component */}
        <Dropdown.Action as='input' placeholder='Enter search' className='w-100' />
        <Dropdown.Content>
          <Dropdown.Item>Interactive Item 1</Dropdown.Item>
          <Dropdown.Item>Interactive Item 2</Dropdown.Item>
        </Dropdown.Content>
      </>
    ),
  },
}
