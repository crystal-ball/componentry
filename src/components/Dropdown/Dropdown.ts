import { activeContainerBuilder } from '../../utils/active-container-component-builder'
import { activeActionBuilder } from '../../utils/active-action-component-builder'
import { activeContentBuilder } from '../../utils/active-content-component-builder'
import {
  ActiveActionBaseProps,
  ActiveContainerBaseProps,
  ActiveContentBaseProps,
  ComponentBaseProps,
} from '../../utils/types'
import { Button } from '../Button/Button'

export interface DropdownProps
  extends ActiveContainerBaseProps,
    ComponentBaseProps<'div'> {}

export interface DropdownActionProps
  extends ActiveActionBaseProps,
    ComponentBaseProps<'button'> {
  /** Display variant */
  variant?: 'primary'
}

export interface DropdownContentProps
  extends ActiveContentBaseProps,
    ComponentBaseProps<'div'> {
  /** Display variant */
  variant?: 'primary'
}

export interface DropdownItemProps
  extends ActiveActionBaseProps,
    ComponentBaseProps<'button'> {
  /** Display variant */
  variant?: 'unstyled'
}

export interface Dropdown {
  (props: DropdownProps): React.ReactElement
  /**
   * [Dropdown action component üìù](https://componentry.design/components/dropdown)
   */
  Action: React.FC<DropdownActionProps>
  /**
   * [Dropdown content component üìù](https://componentry.design/components/dropdown)
   */
  Content: React.FC<DropdownContentProps>
  /**
   * [Dropdown item component üìù](https://componentry.design/components/dropdown)
   */
  Item: React.FC<DropdownItemProps>
}

/**
 * [Dropdown component üìù](https://componentry.design/components/dropdown)
 * @experimental
 */
export const Dropdown = activeContainerBuilder<DropdownProps>('Dropdown', {
  clickEvents: true,
  direction: 'bottom',
  escEvents: true,
}) as Dropdown

Dropdown.Action = activeActionBuilder<DropdownActionProps>('DropdownAction', {
  aria: { expanded: true, haspopup: true, id: true },
  defaultAs: Button,
})

Dropdown.Content = activeContentBuilder<DropdownContentProps>('DropdownContent', {
  aria: { labelledby: true, hidden: true },
})

Dropdown.Item = activeActionBuilder<DropdownItemProps>('DropdownItem')
