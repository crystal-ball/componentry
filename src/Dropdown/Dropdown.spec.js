import React from 'react'
import { render } from '@testing-library/react'

import Dropdown from './Dropdown'
import activationTestSuite from '../../test/activation-tests'
import elementTests from '../../test/element-tests'

describe('<Dropdown />', () => {
  // Basic library activation test suite
  activationTestSuite(Dropdown, {
    name: 'dropdown',
    testArias: ['labelledby', 'expanded'],
  })

  // Basic library element test suite
  elementTests(Dropdown)
  elementTests(Dropdown.Trigger)
  elementTests(Dropdown.Content)
  elementTests(Dropdown.Item)

  it('renders the correct directional classes using direction', () => {
    const { getByTestId, rerender } = render(
      <Dropdown Content='Testing' data-testid='dropdown' Trigger='Toggle' />,
    )

    expect(getByTestId('dropdown')).toHaveClass('dropdown bottom') // default value

    rerender(
      <Dropdown
        Content='Testing'
        data-testid='dropdown'
        direction='top'
        Trigger='Toggle'
      />,
    )

    expect(getByTestId('dropdown')).toHaveClass('dropdown top') // default value
  })
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Dropdown /> snapshots', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Dropdown>
        <Dropdown.Trigger>Trigger</Dropdown.Trigger>
        <Dropdown.Content>
          <Dropdown.Item>Item 1</Dropdown.Item>
          <Dropdown.Item>Item 2</Dropdown.Item>
        </Dropdown.Content>
      </Dropdown>,
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
