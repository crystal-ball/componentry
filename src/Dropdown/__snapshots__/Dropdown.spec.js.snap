// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Drawer /> snapshots renders correctly 1`] = `
<div
  class="dropdown bottom"
  data-id="guid"
>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="dropdown-trigger btn-anchor"
    id="guid"
    type="button"
  >
    Trigger
  </button>
  <div
    aria-hidden="true"
    aria-labelledby="guid"
    class="dropdown-content"
  >
    <button
      class="dropdown-item"
      type="button"
    >
      Item 1
    </button>
    <button
      class="dropdown-item"
      type="button"
    >
      Item 2
    </button>
  </div>
</div>
`;
